File: packages/core/test/passenger_test.dart
import 'dart:convert';

import 'package:mowasalat_core/mowasalat_core.dart';
import 'package:test/test.dart';

import 'fixtures.dart';

/// The real network, as a passenger's phone caches it.
List<Route> _network() => (fixture('routes')['routes'] as List<dynamic>)
    .map((r) => Route.fromJson(r as Map<String, dynamic>))
    .toList(growable: false);

Route _malka() => _network().firstWhere((r) => r.id == 'jo-irbid-malka');

class _FakeBackend {
  final List<HttpRequest> sent = <HttpRequest>[];
  int failNextWith = 0;

  Future<HttpReply> call(HttpRequest request) async {
    sent.add(request);
    if (failNextWith != 0) {
      final status = failNextWith;
      failNextWith = 0;
      return HttpReply(status, jsonEncode({'error': 'refused'}));
    }
    return switch (request.path) {
      '/v1/buses' => HttpReply(200, jsonEncode(fixture('buses'))),
      '/v1/requests' => HttpReply(200, jsonEncode(fixture('ride_request'))),
      '/v1/requests/cancel' => HttpReply(200, jsonEncode({'cancelled': true})),
      '/v1/requests/boarded' => HttpReply(200, jsonEncode({'ok': true})),
      _ => HttpReply(404, jsonEncode({'error': 'not found'})),
    };
  }
}

({PassengerController controller, _FakeBackend backend}) _harness({int Function()? now}) {
  final backend = _FakeBackend();
  return (
    controller: PassengerController(
      api: MowasalatApi(send: backend.call),
      network: _network(),
      remainingBucketM: 250,
      now: now,
    ),
    backend: backend,
  );
}

void main() {
  group('finding a place by what she typed', () {
    test('spelling differences that are not meaning differences still match', () {
      // Each of these is how a real person might type ملكا.
      for (final typed in ['ملكا', 'ملكة', 'مَلكا', 'الملكا', 'ملكـا']) {
        expect(matchesPlace(typed, 'ملكا'), isTrue, reason: 'failed on "$typed"');
      }
    });

    test('a different village does not match', () {
      expect(matchesPlace('أم قيس', 'ملكا'), isFalse);
      expect(matchesPlace('حبراص', 'كفرسوم'), isFalse);
      expect(matchesPlace('سما الروسان', 'كفرسوم'), isFalse);
    });

    test('the typo tolerance does not confuse two real pilot villages', () {
      // Every pair of destinations in the pilot network must stay distinct.
      const villages = ['ملكا', 'سما الروسان', 'كفرسوم', 'حبراص', 'أم قيس'];
      for (final a in villages) {
        for (final b in villages) {
          if (a == b) continue;
          expect(matchesPlace(a, b), isFalse, reason: '"$a" matched "$b"');
        }
      }
    });

    test('folding is idempotent and strips the article', () {
      expect(foldArabic('الملكا'), foldArabic('ملكا'));
      expect(foldArabic(foldArabic('أُمّ قيس')), foldArabic('أُمّ قيس'));
    });

    test('an empty query matches nothing rather than everything', () {
      expect(matchesPlace('', 'ملكا'), isFalse);
      expect(matchesPlace('   ', 'ملكا'), isFalse);
    });

    test('the destination list covers the pilot network', () {
      final destinations = _harness().controller.destinations();
      expect(destinations, hasLength(5));
      expect(destinations.map((d) => d.nameAr), contains('ملكا'));
      expect(destinations.map((d) => d.nameAr), contains('أم قيس'));
    });

    test('search narrows the list', () {
      final found = _harness().controller.destinations(query: 'ملكة');
      expect(found, hasLength(1));
      expect(found.single.id, 'jo-malka');
      expect(found.single.routeIds, contains('jo-irbid-malka'));
    });
  });

  group('which line would take her there', () {
    test('standing at the Irbid end, the line towards Malka is offered', () {
      final h = _harness();
      final irbid = _malka().corridor.zones.first.centre;

      final rides = h.controller.ridesFor(destinationId: 'jo-malka', position: irbid);
      expect(rides, isNotEmpty);
      expect(rides.first.route.id, 'jo-irbid-malka');
      expect(rides.first.dir, 0, reason: 'towards Malka');
    });

    test('standing past her destination is not a ride', () {
      // At Malka itself, going to Malka: nothing is ahead of her.
      final h = _harness();
      final malka = _malka().corridor.zones.last.centre;

      expect(h.controller.ridesFor(destinationId: 'jo-malka', position: malka), isEmpty);
    });

    test('somewhere off every corridor offers nothing', () {
      final h = _harness();
      expect(
        h.controller.ridesFor(destinationId: 'jo-malka', position: const LatLng(31.9, 36.5)),
        isEmpty,
      );
    });

    test('a destination no line serves offers nothing', () {
      final h = _harness();
      final irbid = _malka().corridor.zones.first.centre;
      expect(h.controller.ridesFor(destinationId: 'jo-nowhere', position: irbid), isEmpty);
    });

    test('her own distance is computed locally, never asked for', () {
      final h = _harness();
      final irbid = _malka().corridor.zones.first.centre;
      final ride = h.controller.ridesFor(destinationId: 'jo-malka', position: irbid).first;

      expect(ride.remainingM, greaterThan(5000), reason: 'the line is kilometres long');
      expect(h.backend.sent, isEmpty, reason: 'no request was made to work this out');
    });
  });

  group('seeing the buses', () {
    test('a lookup returns sightings and a stream ticket', () async {
      final h = _harness();
      final irbid = _malka().corridor.zones.first.centre;
      final ride = h.controller.ridesFor(destinationId: 'jo-malka', position: irbid).first;

      final result = await h.controller.busesFor(ride);
      expect(result.streamTicket, isNotEmpty);
      expect(result.buses, isNotEmpty);
      expect(result.buses.first.pseudonym, isNotEmpty);
    });

    test('the position sent is bucketed', () async {
      final h = _harness();
      final irbid = _malka().corridor.zones.first.centre;
      final ride = h.controller.ridesFor(destinationId: 'jo-malka', position: irbid).first;
      await h.controller.busesFor(ride);

      final body = h.backend.sent.single.body!;
      expect((body['remainingM']! as double) % 250, 0);
      expect(jsonEncode(body), isNot(contains('lat')));
    });

    test('the gap and the ETA are worked out on the phone', () {
      // Which is why a passenger watching a stream never sends her position.
      final h = _harness();
      final irbid = _malka().corridor.zones.first.centre;
      final ride = h.controller.ridesFor(destinationId: 'jo-malka', position: irbid).first;

      final sightings = h.controller.sightingsFrom([
        {'pseudonym': 'a', 'remainingM': ride.remainingM + 3000, 'speedKph': 40.0},
      ], ride);

      expect(sightings, hasLength(1));
      expect(sightings.single.gapM, closeTo(3000, 1));
      expect(sightings.single.etaSeconds, closeTo(270, 1));
      expect(sightings.single.etaText, 'بعد 5 دقايق');
    });

    test('a bus that has already passed her is not shown', () {
      final h = _harness();
      final irbid = _malka().corridor.zones.first.centre;
      final ride = h.controller.ridesFor(destinationId: 'jo-malka', position: irbid).first;

      final sightings = h.controller.sightingsFrom([
        {'pseudonym': 'gone', 'remainingM': ride.remainingM - 1000, 'speedKph': 40.0},
      ], ride);
      expect(sightings, isEmpty);
    });

    test('a bus far beyond the window is not shown', () {
      final h = _harness();
      final irbid = _malka().corridor.zones.first.centre;
      final ride = h.controller.ridesFor(destinationId: 'jo-malka', position: irbid).first;

      final sightings = h.controller.sightingsFrom([
        {'pseudonym': 'far', 'remainingM': ride.remainingM + 40000, 'speedKph': 40.0},
      ], ride);
      expect(sightings, isEmpty);
    });

    test('a stopped bus is shown without a made-up arrival time', () {
      final h = _harness();
      final irbid = _malka().corridor.zones.first.centre;
      final ride = h.controller.ridesFor(destinationId: 'jo-malka', position: irbid).first;

      final sightings = h.controller.sightingsFrom([
        {'pseudonym': 'stopped', 'remainingM': ride.remainingM + 2000, 'speedKph': 0.0},
      ], ride);

      expect(sightings.single.etaSeconds, isNull);
      expect(sightings.single.etaText, 'الباص واقف');
    });

    test('the nearest bus is listed first', () {
      final h = _harness();
      final irbid = _malka().corridor.zones.first.centre;
      final ride = h.controller.ridesFor(destinationId: 'jo-malka', position: irbid).first;

      final sightings = h.controller.sightingsFrom([
        {'pseudonym': 'far', 'remainingM': ride.remainingM + 8000, 'speedKph': 40.0},
        {'pseudonym': 'near', 'remainingM': ride.remainingM + 1000, 'speedKph': 40.0},
      ], ride);

      expect(sightings.map((s) => s.pseudonym), ['near', 'far']);
    });

    test('a real stream snapshot parses into sightings', () {
      final h = _harness();
      final irbid = _malka().corridor.zones.first.centre;
      final ride = h.controller.ridesFor(destinationId: 'jo-malka', position: irbid).first;

      final buses = (fixture('stream_buses')['buses'] as List<dynamic>)
          .cast<Map<String, dynamic>>();
      // Whether it is ahead of her depends on where she stands; what matters is
      // that a captured payload feeds straight through without reshaping.
      expect(() => h.controller.sightingsFrom(buses, ride), returnsNormally);
    });
  });

  group('waiting', () {
    test('أنا مستني هون records a request and nothing about her', () async {
      final h = _harness();
      final irbid = _malka().corridor.zones.first.centre;
      final ride = h.controller.ridesFor(destinationId: 'jo-malka', position: irbid).first;

      final pseudonym = await h.controller.requestRide(ride);
      expect(pseudonym, isNotEmpty);
      expect(h.controller.state, WaitState.waiting);

      final body = h.backend.sent.last.body!;
      expect(body.keys.toSet(), {'routeId', 'dir', 'destinationId', 'remainingM', 'zoneSeq'});
      expect(h.backend.sent.last.bearer, isNull, reason: 'she has no account at all');
    });

    test('she cannot be waiting twice', () async {
      final h = _harness();
      final irbid = _malka().corridor.zones.first.centre;
      final ride = h.controller.ridesFor(destinationId: 'jo-malka', position: irbid).first;

      await h.controller.requestRide(ride);
      await expectLater(h.controller.requestRide(ride), throwsStateError);
    });

    test('the request expires on its own, as the backend drops it', () async {
      var clock = 1000;
      final h = _harness(now: () => clock);
      final irbid = _malka().corridor.zones.first.centre;
      final ride = h.controller.ridesFor(destinationId: 'jo-malka', position: irbid).first;

      await h.controller.requestRide(ride);
      expect(h.controller.timeLeft.inMinutes, 20);

      clock += PassengerController.requestLifetime.inMilliseconds + 1;
      expect(h.controller.hasExpired, isTrue);
      expect(h.controller.refresh(), WaitState.expired);
      expect(h.controller.timeLeft, Duration.zero);
    });

    test('إلغاء clears it here and there', () async {
      final h = _harness();
      final irbid = _malka().corridor.zones.first.centre;
      final ride = h.controller.ridesFor(destinationId: 'jo-malka', position: irbid).first;

      await h.controller.requestRide(ride);
      await h.controller.cancel();

      expect(h.controller.state, WaitState.cancelled);
      expect(h.controller.requestPseudonym, isNull);
      expect(h.backend.sent.map((r) => r.path), contains('/v1/requests/cancel'));
    });

    test('ركبت closes the request', () async {
      final h = _harness();
      final irbid = _malka().corridor.zones.first.centre;
      final ride = h.controller.ridesFor(destinationId: 'jo-malka', position: irbid).first;

      await h.controller.requestRide(ride);
      await h.controller.boarded();

      expect(h.controller.state, WaitState.boarded);
      expect(h.controller.chosenRide, isNull);
      expect(h.backend.sent.map((r) => r.path), contains('/v1/requests/boarded'));
    });

    test('a backend that has already forgotten her still clears locally', () async {
      final h = _harness();
      final irbid = _malka().corridor.zones.first.centre;
      final ride = h.controller.ridesFor(destinationId: 'jo-malka', position: irbid).first;

      await h.controller.requestRide(ride);
      h.backend.failNextWith = 400;
      await h.controller.cancel();

      expect(h.controller.state, WaitState.cancelled);
      expect(h.controller.requestPseudonym, isNull);
    });

    test('cancelling when she never asked is harmless', () async {
      final h = _harness();
      await h.controller.cancel();
      expect(h.controller.state, WaitState.idle);
    });

    test('reset returns her to وين رايح؟', () async {
      final h = _harness();
      final irbid = _malka().corridor.zones.first.centre;
      final ride = h.controller.ridesFor(destinationId: 'jo-malka', position: irbid).first;

      await h.controller.requestRide(ride);
      await h.controller.boarded();
      h.controller.reset();

      expect(h.controller.state, WaitState.idle);
    });
  });
}

File: apps/web-passenger/test/passenger.test.ts
import { test } from "node:test";
import assert from "node:assert/strict";
import { readFileSync } from "node:fs";
import { join } from "node:path";

import { PassengerController, etaText, REQUEST_LIFETIME_MS } from "../src/passenger.ts";
import type { Api, BusScalar } from "../src/api.ts";
import type { Direction, LatLng, Route } from "../../../packages/corridor/src/types.ts";

/**
 * The same recordings the Dart client is tested against, captured from a live
 * /v1 backend. Testing both clients on one set of responses is what keeps them
 * from drifting into two different ideas of the same journey.
 */
const fixtures = join(import.meta.dirname, "../../../packages/core/test/fixtures");

function load<T>(name: string): T {
  return JSON.parse(readFileSync(join(fixtures, name), "utf8")) as T;
}

const network = load<{ routes: Route[] }>("routes.json").routes;
const BUCKET_M = load<{ remainingBucketM: number }>("country.json").remainingBucketM;

const IRBID: LatLng = { lat: 32.5556, lng: 35.8497 };
const MALKA: LatLng = { lat: 32.669, lng: 35.744 };

/** Records what was sent, so a test can assert on the wire and not just the result. */
function recorder(overrides: Partial<Api> = {}) {
  const sent: { path: string; body: Record<string, unknown> }[] = [];
  const api = {
    async findBuses(args: Record<string, unknown>) {
      sent.push({ path: "/v1/buses", body: args });
      return { buses: [] as BusScalar[], streamTicket: "t-1" };
    },
    async requestRide(args: Record<string, unknown>) {
      sent.push({ path: "/v1/requests", body: args });
      return { pseudonym: "p-1" };
    },
    async cancelRequest(pseudonym: string) {
      sent.push({ path: "/v1/requests/cancel", body: { pseudonym } });
      return { cancelled: true };
    },
    async boarded(args: Record<string, unknown>) {
      sent.push({ path: "/v1/requests/boarded", body: args });
      return { ok: true as const };
    },
    ...overrides,
  } as unknown as Api;
  return { api, sent };
}

function controllerAt(now = () => 0) {
  const { api, sent } = recorder();
  return { c: new PassengerController(api, network, BUCKET_M, now), sent };
}

// --- where can she go -------------------------------------------------------

test("every served place is offered once, whichever lines reach it", () => {
  const { c } = controllerAt();
  const options = c.destinations();

  assert.equal(options.length, new Set(options.map((o) => o.id)).size);
  assert.ok(options.some((o) => o.id === "jo-malka"));
  for (const o of options) assert.ok(o.routeIds.length > 0);
});

test("search finds a village through a misspelling", () => {
  const { c } = controllerAt();
  assert.deepEqual(
    c.destinations("ملكة").map((o) => o.id),
    ["jo-malka"],
  );
});

// --- which line would pick her up -------------------------------------------

test("standing at the hub, the line towards her destination is offered", () => {
  const { c } = controllerAt();
  const rides = c.ridesFor("jo-malka", IRBID);

  assert.equal(rides.length, 1);
  assert.equal(rides[0].route.id, "jo-irbid-malka");
  assert.equal(rides[0].dir, 0, "origin → destination");
  assert.ok(rides[0].remainingM > 0);
});

test("standing at the destination is not a ride, it is a walk back", () => {
  const { c } = controllerAt();
  // She is already in ملكا and asks for ملكا. Direction 0 has nothing left to
  // cover, and direction 1 runs away from it.
  assert.deepEqual(c.ridesFor("jo-malka", MALKA), []);
});

test("a position nowhere near any corridor offers nothing", () => {
  const { c } = controllerAt();
  assert.deepEqual(c.ridesFor("jo-malka", { lat: 31.95, lng: 35.93 }), []); // Amman
});

// --- the wire ---------------------------------------------------------------

test("what is sent is bucketed, and carries no coordinate", () => {
  const { c, sent } = controllerAt();
  const ride = c.ridesFor("jo-malka", IRBID)[0];

  return c.busesFor(ride).then(async () => {
    await c.requestRide(ride);

    assert.equal(sent.length, 2);
    for (const { body } of sent) {
      assert.equal(
        (body.remainingM as number) % BUCKET_M,
        0,
        "a position finer than the country's band would be refused by the server",
      );
      for (const key of Object.keys(body)) {
        assert.ok(
          !/(^|_)(lat|latitude|lng|lon|longitude|coords?|position|location)(_|$)/i.test(key),
          `${key} must never be sent`,
        );
      }
    }
  });
});

// --- watching the bus -------------------------------------------------------

test("only buses still behind her, and only within the window", () => {
  const { c } = controllerAt();
  const ride = c.ridesFor("jo-malka", IRBID)[0];

  const sightings = c.sightingsFrom(
    [
      { pseudonym: "ahead", remainingM: ride.remainingM - 1000, zoneSeq: 0, speedKph: 40 },
      { pseudonym: "behind", remainingM: ride.remainingM + 2000, zoneSeq: 0, speedKph: 40 },
      { pseudonym: "far", remainingM: ride.remainingM + 40_000, zoneSeq: 0, speedKph: 40 },
    ],
    ride,
  );

  assert.deepEqual(
    sightings.map((s) => s.pseudonym),
    ["behind"],
  );
  assert.equal(sightings[0].gapM, 2000);
});

test("the nearest bus is listed first", () => {
  const { c } = controllerAt();
  const ride = c.ridesFor("jo-malka", IRBID)[0];

  const sightings = c.sightingsFrom(
    [
      { pseudonym: "further", remainingM: ride.remainingM + 5000, zoneSeq: 0, speedKph: 40 },
      { pseudonym: "nearer", remainingM: ride.remainingM + 1000, zoneSeq: 0, speedKph: 40 },
    ],
    ride,
  );

  assert.deepEqual(
    sightings.map((s) => s.pseudonym),
    ["nearer", "further"],
  );
});

test("a stopped bus gets no ETA, because a number would be a guess", () => {
  const { c } = controllerAt();
  const ride = c.ridesFor("jo-malka", IRBID)[0];

  const [stopped] = c.sightingsFrom(
    [{ pseudonym: "x", remainingM: ride.remainingM + 3000, zoneSeq: 0, speedKph: 0 }],
    ride,
  );

  assert.equal(stopped.etaSeconds, null);
  assert.equal(etaText(stopped), "الباص واقف");
});

test("an ETA is distance over speed, read as minutes", () => {
  const { c } = controllerAt();
  const ride = c.ridesFor("jo-malka", IRBID)[0];

  // 6 km at 36 km/h is ten minutes.
  const [bus] = c.sightingsFrom(
    [{ pseudonym: "x", remainingM: ride.remainingM + 6000, zoneSeq: 0, speedKph: 36 }],
    ride,
  );

  assert.equal(Math.round(bus.etaSeconds!), 600);
  assert.equal(etaText(bus), "بعد 10 دقايق");
});

// --- the one interaction she has --------------------------------------------

test("waiting, then boarding, leaves nothing behind", async () => {
  const { c, sent } = controllerAt();
  const ride = c.ridesFor("jo-malka", IRBID)[0];

  assert.equal(c.state, "idle");
  await c.requestRide(ride);
  assert.equal(c.state, "waiting");
  assert.equal(c.requestPseudonym, "p-1");

  await c.boarded();
  assert.equal(c.state, "boarded");
  assert.equal(c.requestPseudonym, null);
  assert.equal(c.chosenRide, null);
  assert.ok(sent.some((s) => s.path === "/v1/requests/boarded"));
});

test("cancelling clears the local state even when the server has already forgotten", async () => {
  const { api } = recorder({
    async cancelRequest() {
      throw new Error("gone");
    },
  } as Partial<Api>);
  const c = new PassengerController(api, network, BUCKET_M, () => 0);
  const ride = c.ridesFor("jo-malka", IRBID)[0];

  await c.requestRide(ride);
  await c.cancel();

  assert.equal(c.state, "cancelled");
  assert.equal(c.requestPseudonym, null);
});

test("the request expires on its own, exactly when the server drops it", async () => {
  let now = 0;
  const { c } = controllerAt(() => now);
  const ride = c.ridesFor("jo-malka", IRBID)[0];

  await c.requestRide(ride);
  assert.equal(c.timeLeftMs, REQUEST_LIFETIME_MS);

  now = REQUEST_LIFETIME_MS - 1000;
  assert.equal(c.refresh(), "waiting");
  assert.equal(c.timeLeftMs, 1000);

  now = REQUEST_LIFETIME_MS;
  assert.equal(c.refresh(), "expired");
  assert.equal(c.timeLeftMs, 0);
});

test("she cannot be waiting for two buses at once", async () => {
  const { c } = controllerAt();
  const ride = c.ridesFor("jo-malka", IRBID)[0];

  await c.requestRide(ride);
  await assert.rejects(() => c.requestRide(ride));
});

test("a server that did not say how coarse positions must be is refused", () => {
  // She is only protected because everything is rounded to the country's band.
  // A missing band is a misconfiguration, not a default to invent.
  const { api } = recorder();
  for (const bad of [undefined, 0, -250, Number.NaN]) {
    assert.throws(
      () => new PassengerController(api, network, bad as unknown as number),
      /remainingBucketM/,
    );
  }
});
